### ISSUE TITLE
2020/3/1

### ISSUE THREAD

**Issue by mingxoxo** [mingxoxo/deapLearning-study#5]
Deeper Look at Gradient Descent

Hypothesis(Linear Regression)
인공신경망의 구조를 나타낸다.
W는 하나의 matrix
![image](https://user-images.githubusercontent.com/44529556/75621731-73fac080-5bdb-11ea-9bd4-08faddb76087.png)

Simpler Hypothesis Function (No Bias)
![image](https://user-images.githubusercontent.com/44529556/75621738-8ecd3500-5bdb-11ea-8cad-5bc2037a0263.png)

![image](https://user-images.githubusercontent.com/44529556/75621741-9d1b5100-5bdb-11ea-857c-2693e54003b8.png)

![image](https://user-images.githubusercontent.com/44529556/75621748-b1f7e480-5bdb-11ea-809a-18d78fbbdb0b.png)
W가 1에 수렴할수록 더 정확한 모델

**Cost function: Intuition**
 모델의 예측값이 실제 데이터와 얼마나 다른지 나타내는 값
잘 학습된 모델일수록 낮은 cost를 가진다.
W = 1일때 cost = 0
![image](https://user-images.githubusercontent.com/44529556/75621775-2f235980-5bdc-11ea-9e29-ebb61264461f.png)

**MSE**
예측값과 실제값의 차이를 제곱한 평균을 구해준다.
![image](https://user-images.githubusercontent.com/44529556/75621780-4f531880-5bdc-11ea-8312-5f51bd88f36d.png)

**Gradient Descent: Intuition**
우리가 원하는 것은 cost function을 최소화시키는 일
-> 기울기가 양수일때는 W가 더 작아져야하고 음수일때는 W가 더 커져야 한다.
-> 기울기가 가파를수록 cost가 큰것이니 W를 크게 바꾸고
기울기가 평평할수록 cost가 작은것이니 W를 작게 바꾸어야 한다.
우리는 이 기울기를 Gradient라고 한다.
![image](https://user-images.githubusercontent.com/44529556/75621815-c7b9d980-5bdc-11ea-948b-4a46dc975466.png)
(역삼각형 W 가 정확히는 cost 함수를 W로 미분한 값입니다.
역삼각형인 이유는 partial derivative 이기때문입니다..?)

![image](https://user-images.githubusercontent.com/44529556/75621856-2da66100-5bdd-11ea-94b2-1482b2923073.png)

**Comment by mingxoxo** [comment]
**Full Code**

```
import torch
#데이터
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[1], [2], [3]])
#모델 초기화
W = torch.zeros(1)
#learning rate 설정
lr = 0.1
#Epoch : 데이터로 학습한 횟수
nb_epochs = 10
for epoch in range(nb_epochs + 1) :

  #H(x) 계산
  hypothesis = x_train * W

  #cost gradient 계산
  cost = torch.mean((hypothesis - y_train) ** 2)
  gradient = torch.sum((W * x_train - y_train)*x_train)

  print('Epoch {:4d}/{} W: {:.3f}, Cost: {:.6f}'.format(epoch, nb_epochs, W.item(), cost.item()))

  #cost gradient로 H(x) 개선
  W -= lr * gradient
```
